In [52]:
# Descriptive Stats: Thống kê mô tả
# Có thực sự có mối liên hệ không, hay do ngẫu nhiên mà có
  # Statistical Significance (Ý nghĩa về mặt thống kê hay không)
  # Inferential Statistics: thống kê suy luận -> Hypothesis Testing: kiểm định giả thuyết

# Hypothesis Testing: Kiểm định giả thuyết
# Đặt ra giả thuyết
  # Null hypothesis (Giả thuyết không)
  # Alternative hypothesis (Giả thuyết thay thế)
  # Để chứng minh được giả thuyết thay thế là đúng, chúng ta cần bác bỏ được giả thuyết không
  # Trong trường hợp không bác bỏ được giả thuyết không,
    # chúng ta chưa để căn cứ để chứng minh giả thuyết thay thế là đúng

# Kiểm định giả thuyết về giá trị trung bình:
  # Ban đầu, điểm thi là 60 điểm.
    # Sau khi áp dụng cách dạy mới, điểm trung bình của học sinh lớp phân tích dữ liệu có thay đổi không?
      # H0: = 60
      # H1: khác 60
      # có thể nhỏ hơn 60, có thể lớn hơn 60 -> 2 tail hypothesis
    # Sau khi áp dụng cách dạy mới, điểm trung bình có tăng lên không
      # H0: <=60
      # H1: >60
    # Sau khi áp dụng cách dạy mới, điểm trung bình có bị giảm đi không
      # H0: >=60
      # H1: <60
# Kiểm định giả thuyết so sánh giá trị trung bình:
  # Có 2 logo, cần chọn xem nên dùng logo nào cho trang facebook quảng cáo.
  # Tiêu chí để chọn xem nên dùng logo nào dựa vào số lượt click trung bình cho mỗi logo. (A/B Testing)
  # Số lượt click của hai logo có khác nhau không?
    # H0: x1-x2=0
    # H1: x1-x2 khác 0
  # Số lượt click của logo 1 có nhiều hơn logo 2 không?
    # H0: x1-x2 <= 0
    # H1: x1-x2 > 0
# Kiểm định giả thuyết so sánh cặp giá trị:
  # Tiên có cửa hàng bán quần áo. Tiên muốn chạy quảng cáo cho cửa hàng, dự định là tiêu $10,000 để chạy quảng cáo.
  # Mong muốn: sau khi chạy quảng cáo, doanh thu đem về sẽ nhiều hơn doanh thu ban đầu $60,000 một tháng
  # H0: x2 - x1 <= 60000
  # H1: x2 - x1 > 60000

# Sau khi đặt ra các giả thuyết, xem có bác bỏ được Null Hypothesis không?
  # Cách hiểu p-value
    # Nhìn vào p-value: xác suất Null Hypothesis đúng
    # Muốn bác bỏ được Null Hypothesis, p-value phải càng nhỏ càng tốt
    # p-value < 0.05 -> bác bỏ Null Hypothesis
  # Cách tính p-value:
    # Trong trường hợp có đầy đủ dữ liệu của dân số (population) -> Z-statistics (Z-score)
    # Đa phần chúng ta chỉ có dữ liệu mẫu -> T-statistics (ttest)



In [53]:
# Tiếp tục thực hành
# Quay lại bài Quảng Cáo

# Mô tả dữ liệu (phân tích dữ liệu) (Descriptive Stats) -> cho biết tình hình hiện tại của doanh nghiệp ra sao
  # -> Data Analyst, BI Analyst
# Dựa trên tình hình hiện tại (cả quá khứ), dự đoán tương lai (Predictive Stats)
  # Dự đoán doanh thu dựa trên số tiền tiêu cho quảng cáo
  # -> Data Scientist
# Dự đoán được là nếu tiêu 10,000 đô vào quảng cáo trên TV, thì doanh số sẽ tăng lên thành bán được 60,000 sản phẩm
  # Ước tính -> thử mô hình (lấy mẫu) lên 10 cửa hàng -> kiểm định giả thuyết
  # Chạy trong 3 tháng -> Sau 3 tháng chạy mô hình (sau 3 tháng chạy quảng cáo), doanh thu có tăng lên 60,000 sản phẩm không?
  # -> Đưa ra giả pháp và kiểm tra (Prescriptive Statistics)

# => Làm hết cả 3: Manager

# [Thực hành] phân tích dữ liệu trên bộ dữ liệu Quản lý chất lượng

In [54]:
import pandas as pd
from scipy import stats
import matplotlib.pyplot as plt

In [55]:
df = pd.read_excel('Quality.xlsx')
df.head()

C:\Users\Admin\AppData\Local\Programs\Python\Python311\Lib\site-packages\openpyxl\worksheet\header_footer.py:48: UserWarning: Cannot parse header or footer so it will be ignored
  warn("""Cannot parse header or footer so it will be ignored""")


Sample 1  Sample 2  Sample 3  Sample 4
0     11.55     11.62     11.91     12.02
1     11.62     11.69     11.36     12.02
2     11.52     11.59     11.75     12.05
3     11.75     11.82     11.95     12.18
4     11.90     11.97     12.14     12.11

Quality Associates là một công ty tư vấn cho khách hàng về việc lấy mẫu và quản lý chất lượng cho quá trình sản xuất. Một khách hàng gửi cho Quality Associates một bộ dữ liệu chứa thông tin của 800 mẫu sản xuất trong điều kiện nhà máy hoạt động tốt và khách hàng cho rằng trọng lượng trung bình của mỗi gói hàng là 12, và độ lệch không quá 0.21. Để kiểm tra lại, Quality Associates gợi ý khách hàng rằng sẽ lấy ngẫu nhiên các 4 mẫu dữ liệu mới, mỗi mẫu chứa 30 sản phẩm. Để kiểm định rằng hoạt động của nhà máy của khách hàng có tốt như những gì họ nghĩ hay không, đại diện cho Quality Associates bạn hãy thực hiện những phân tích kiểm định chất lượng giúp khách hàng. Dữ liệu được để trong file Quality.xlsx

Yêu cầu: Tiến hành viết giả thuyết kiểm định về chất lượng sản phẩm với mức ý nghĩa 0.01

In [56]:
# 4 mẫu dữ liệu mới, mỗi mẫu chứa 30 sản phẩm
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Sample 1  30 non-null     float64
 1   Sample 2  30 non-null     float64
 2   Sample 3  30 non-null     float64
 3   Sample 4  30 non-null     float64
dtypes: float64(4)
memory usage: 1.1 KB


In [57]:
# Kiểm định chất lượng sản phẩm với mức ý nghĩa 0.01
# p_value < 0.01
# Thế nào là chất lượng? -> kiểm tra xem có đúng là trọng lượng mỗi gói hàng là 12 thật không?

In [58]:
df.describe()

Sample 1   Sample 2   Sample 3   Sample 4
count  30.000000  30.000000  30.000000  30.000000
mean   11.958667  12.028667  11.889000  12.081333
std     0.220356   0.220356   0.207171   0.206109
min    11.520000  11.590000  11.360000  11.640000
25%    11.812500  11.882500  11.752500  11.982500
50%    11.955000  12.025000  11.920000  12.080000
75%    12.145000  12.215000  11.997500  12.227500
max    12.320000  12.390000  12.220000  12.470000

In [59]:
# 11.95, 12.02, 11.88, 12.08 -> có thể có
# kiểm tra trọng lượng trung bình có bằng 12 thật không
# nói cách khác, kiểm tra xem 11.95, 12.02... có thực sự lệch khỏi 12 không?
  # lệch đấy có đáng kể về mặt thống kê hay không?

In [60]:
# H0:=12
# H1:khác 12

In [61]:
# chuẩn bị dữ liệu
sample=pd.concat([df['Sample 1'], df['Sample 2'], df['Sample 3'], df['Sample 4']],).reset_index(drop=True)
sample

0      11.55
1      11.62
2      11.52
3      11.75
4      11.90
       ...  
115    12.17
116    11.94
117    11.97
118    12.23
119    12.25
Length: 120, dtype: float64

In [62]:
sample.describe()

count    120.000000
mean      11.989417
std        0.223108
min       11.360000
25%       11.850000
50%       12.000000
75%       12.160000
max       12.470000
dtype: float64

#### Dựa vào kết quả phân tích trên: bộ dữ liệu chứa đủ 120 mẫu

Giá trị trung bình trên mẫu đúng bằng 12, giống mô tả của khách hàng

Phương sai: = 0.223108 (sai số so với giá trị trung bình của mẫu dữ liệu) >0.21 
–> khách hàng nên thay đổi lại tuyên bố về sai số của mình

Tiến hành kiểm định về giá trị trung bình

Giả thuyết không: khối lượng trung bình của sản phẩm = 12
Giả thuyết đối: Khối lượng trung bình của sản phẩm # 12

Thực hiện phép kiểm định: One Sample T Test

In [63]:
stats.ttest_1samp(sample,12)

TtestResult(statistic=-0.5196355789314843, pvalue=0.6042828222996104, df=119)

In [64]:
# pvalue = 0.6 >>> 0.01
# Chưa đủ cơ sở để bác bỏ giả thuyết khối lượng = 12
# Chưa đủ cơ sở để chứng minh có sự khác biệt về khối lượng mẫu trung bình so với tiêu chuẩn đặt ra

#### Giải thích trên trang learn cuả codegym

Do pvalue > 0.01 –> Chấp nhận giả thuyết 0.

Kết luận: Tuyên bố của khách hàng về giá trị trung bình là đúng, nhưng tuyên bố về sai số của sản phẩm đang không hợp lý.

Khách hàng nên tuyên bố như sau: Khối lượng sản phẩm trung bình của sản phẩm là 12, với sai số đóng gói trung bình là 0.24 (chiếm khoảng 2

%). Hoặc khách hàng cần phải xem lại việc vận hành để đảm bảo tuyên bố ban đầu.

# [Thực hành] Kiểm định lượng sản phẩm tiêu thụ của Coca Cola

### Mục tiêu
So sánh giá trị trung bình giữa hai tổng thể

Cách viết kiểm định giả thuyết

Cách kết luận cho 1 giả thuyết thống kê

Thực hiện kiểm định Independent T test

In [75]:
df1 = pd.read_excel("Coca_cola_use.xlsx", index_col= 'STT')
df1.head()

Ohio  Atlanta
STT               
1    0.88     3.85
2    2.06     0.00
3    0.00     0.00
4    0.70     0.02
5    1.64     0.58

Bộ phận bán hàng của Công ty Coca Cola tuyên bố rằng lượng tiêu thụ coca trung bình ở Ohio lớn hơn ở Atlanta. 

Để tiến hành chiến lược phân phối sản phẩm ở 2 khu vực này, bộ phận Marketing có thu thập số liệu sử dụng coca của 50 người ở Ohio và 50 người ở Atlanta. 

Dữ liệu được để trong file Coca_cola_use.xlsx, với đơn vị tiêu thụ được tính l/năm cho 1 người. 

Với mức ý nghĩa 5% hãy kiểm định giả thuyết xem có đúng lượng tiêu thụ Coca bình quân ở Ohio lớn hơn Atlanta hay không

In [66]:
# Giả thuyết: lượng tiêu thụ coca trung bình Ohio > Atlanta
# 50 dòng

In [76]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50 entries, 1 to 50
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Ohio     50 non-null     float64
 1   Atlanta  50 non-null     float64
dtypes: float64(2)
memory usage: 1.2 KB


In [77]:
df1.describe()

Ohio    Atlanta
count  50.000000  50.000000
mean    1.746800   1.614800
std     1.092211   1.061068
min     0.000000   0.000000
25%     0.835000   0.682500
50%     1.615000   1.495000
75%     2.350000   2.402500
max     5.000000   3.860000

In [79]:
df1.boxplot()

<Axes: >

nhìn vào biểu đồ trên chúng ta thấy lượng tiêu thụ ở Ohio có nhỉnh hơn Atlanta một chút, và rất khó để kết luận

#### Giả thuyết của cô Tiên 
H0: Ohio - Atlanta <= 0

H1: Ohio - Atlanta > 0

#### Tiến hành kiểm định - codegym

Thực hiện kiểm định giả thuyết so sánh hai mẫu trung bình độc lập gọi a1, a2 lần lượt là lượng tiêu thụ coca trung bình trên đầu người ở Ohio và Atlanta

Giả thuyết không: a1-a2 =0

Giả thuyết đối: a1-a2>0

mức ý nghĩa 5%

Loại kiểm định Independent T test

In [69]:
stats.ttest_ind(df1['Ohio'], df1['Atlanta'], equal_var=False) # khác nhau

TtestResult(statistic=0.6129544104328379, pvalue=0.5413275550760535, df=97.91810815759084)

pvalue=0.541 > 0.05

Chúng ta nhìn thấy rằng: pvalue > 5% rất nhiều nên không đủ cơ sở để bác bỏ giả thuyết không 

Kết luận: Không đủ căn cứ để kết luận rằng lượng tiêu thụ coca trung bình ở Ohio lớn hơn ở Atlanta

In [70]:
stats.ttest_ind(df['Ohio'], df['Atlanta'], equal_var=False, alternative='greater') # lớn hơn

TtestResult(statistic=0.6129544104328379, pvalue=0.27066377753802673, df=97.91810815759084)

In [71]:
stats.ttest_ind(df['Ohio'], df['Atlanta'], equal_var=False, alternative='less') # nhỏ hơn

TtestResult(statistic=0.6129544104328379, pvalue=0.7293362224619733, df=97.91810815759084)

In [72]:
# Chưa đủ bằng chứng để nói rằng số lượng tiêu thụ Coca ở Ohio lớn hơn ở Atlanta